### PR TITLE
[e2e] make test bindings friendlier to integration tests

### PR DESCRIPTION
**Pull request by dnfield**
## Description

Integration tests may not want to override testTextInput or HttpClient. This lets binding subclasses for tests achieve that.

## Related Issues

Part of https://github.com/flutter/flutter/issues/51885
Part of https://github.com/flutter/flutter/issues/57623

## Tests

I added the following tests:

Test that the default bindings _do_ override HttpClient and register TestTextInput.
Test that custom bindings using the new properties do not override/register.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter_test/lib/src/binding.dart [hunk 1]
```diff
@@ -269,8 +290,13 @@ abstract class TestWidgetsFlutterBinding extends BindingBase
   void initInstances() {
     super.initInstances();
     timeDilation = 1.0; // just in case the developer has artificially changed it for development
-    binding.setupHttpOverrides();
-    _testTextInput = TestTextInput(onCleared: _resetFocusedEditable)..register();
+    if (overrideHttpClient) {
+      binding.setupHttpOverrides();
+    }
+    _testTextInput = TestTextInput(onCleared: _resetFocusedEditable);
```
[the remark this answers is not in the document]
Comment (dnfield) [reply packages/flutter_test/lib/src/binding.dart:296]: And I've updated TestTextInput in another patch to assert that it's registered for any method that wouldn't make sense in an unregistered context.

[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/flutter_test/lib/src/binding.dart:296]: sgtm

**Review by chunhtai — APPROVED** [review]
LGTM

**Comment by nturgut** [comment]
Thanks for the change, I was still looking at the PR. LGTM!

**Pull request merged**